File: 4. Count Occurrences of Anagrams.cpp
// Topic: Count Occurrences of Anagrams
#include <bits/stdc++.h>
using namespace std;

int main()
{
    string s = "cbaebabacd";
    string p = "abc";
    int n = s.size(), m = p.size();
    if (m > n)
    {
        cout << 0 << "\n";
        return 0;
    }
    vector<int> ps(26, 0), ws(26, 0);
    for (char c : p) ps[c - 'a']++;
    for (int i = 0; i < m; i++) ws[s[i] - 'a']++;
    int count = 0;
    if (ps == ws)
        count++;
    for (int i = m; i < n; i++)
    {
        ws[s[i] - 'a']++;
        ws[s[i - m] - 'a']--;
        if (ps == ws)
            count++;
    }
    cout << "Anagram occurrences: " << count << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(26) ~ O(1)

File: 4. Count Occurrences of Anagrams.py
"""
Count Occurrences of Anagrams
"""
from collections import Counter

def count_anagrams(s, p):
    n, m = len(s), len(p)
    if m>n: return 0
    pc = Counter(p)
    window = Counter(s[:m])
    count = 1 if window==pc else 0
    for i in range(m, n):
        window[s[i]] += 1
        window[s[i-m]] -= 1
        if window[s[i-m]] == 0: del window[s[i-m]]
        if window==pc: count += 1
    return count

def main():
    print(count_anagrams('cbaebabacd','abc'))

if __name__ == '__main__':
    main()

# Time Complexity: O(n)
# Space Complexity: O(1) (alphabet size)
